addons/payment_alipay/models/payment_transaction.py

```diff
@@ -0,0 +1,157 @@
+                )
+            )
+
+    # === BUSINESS METHODS ===#
+
+    def _get_specific_rendering_values(self, processing_values):
+        if self.acquirer_id.provider != 'alipay':
+            return super()._get_specific_rendering_values(processing_values)
+
+        base_url = self.acquirer_id._get_base_url()
+        alipay_tx_values = {
+            '_input_charset': 'utf-8',
+            'notify_url': urls.url_join(base_url, AlipayController._notify_url),
+            'out_trade_no': processing_values['reference'],
+            'partner': self.acquirer_id.alipay_merchant_partner_id,
+            'return_url': urls.url_join(base_url, AlipayController._return_url),
+            'subject': processing_values['reference'],
+            'total_fee': processing_values['amount'] + self.fees
+        }
+        if self.acquirer_id.alipay_payment_method == 'standard_checkout':
+            # https://global.alipay.com/docs/ac/global/create_forex_trade
+            alipay_tx_values.update({
+                'service': 'create_forex_trade',
+                'product_code': 'NEW_OVERSEAS_SELLER',
+                'currency': self.currency_id.name,
+            })
+        else:
+            alipay_tx_values.update({
+                'service': 'create_direct_pay_by_user',
+                'payment_type': 1,
+                'seller_email': self.acquirer_id.alipay_seller_email,
+            })
+
+        sign = self.acquirer_id._alipay_build_sign(alipay_tx_values)
+        alipay_tx_values.update({
+            'sign_type': 'MD5',
+            'sign': sign,
+            'tx_url': self.acquirer_id._alipay_get_api_url(),
+        })
+        return alipay_tx_values
+
+    @api.model
+    def _get_tx_from_feedback_data(self, provider, data):
+        if provider != 'alipay':
+            return super()._get_tx_from_feedback_data(provider, data)
+
+        reference = data.get('reference') or data.get('out_trade_no')
+        tx = self.env['payment.transaction'].search([('reference', '=', reference)])
+        if not tx:
+            raise ValidationError(
+                "Alipay: " + _(
+                    "Received data with reference %s matching no transaction.", reference
+                )
+            )
+
+        # Verify signature (done here because we need the reference to get the acquirer)
+        sign_check = tx.acquirer_id._alipay_build_sign(data)
+        sign = data.get('sign')
+        if sign != sign_check:
+            raise ValidationError(
+                "Alipay: " + _(
+                    "Expected signature %(sc) but received %(sign)s.", sc=sign_check, sign=sign
+                )
+            )
+        txn_id = data.get('trade_no')
+        if not reference or not txn_id:
+            raise ValidationError(
+                "Alipay: " + _(
+                    "Received data with missing reference %(r)s or txn_id %(t)s.",
+                    r=reference, t=txn_id
+                )
+            )
```

Comment:
Should be bumped before the search (or remove the reference remark, because it won't ever happen, you always have a reference at this step).